overlord/devicestate/handlers.go

```diff
@@ -698,3 +703,231 @@ func fetchKeys(st *state.State, keyID string) (errAcctKey error, err error) {
 		keyID = a.SignKeyID()
 	}
 }
+
+func snapState(st *state.State, name string) (*snapstate.SnapState, error) {
+	var snapst snapstate.SnapState
+	err := snapstate.Get(st, name, &snapst)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	return &snapst, nil
+}
+
+func makeRollbackDir(name string) (string, error) {
+	snapRollbackDir := filepath.Join(dirs.SnapRollbackDir, name)
+
+	if err := os.MkdirAll(dirs.SnapRollbackDir, 0755); err != nil {
+		return "", err
+	}
+
+	if err := os.Mkdir(snapRollbackDir, 0755); err != nil {
+		// fail even if directory exists already
+		return "", err
+	}
+
+	return snapRollbackDir, nil
+}
+
+func currentGadgetInfo(snapst *snapstate.SnapState) (*gadget.Info, error) {
+	var gi *gadget.Info
+
+	currentInfo, err := snapst.CurrentInfo()
+	if err != nil && err != snapstate.ErrNoCurrent {
+		return nil, err
+	}
+	if currentInfo != nil {
+		gi, err = snap.ReadGadgetInfo(currentInfo, release.OnClassic)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return gi, nil
+}
+
+func pendingGadgetInfo(snapsup *snapstate.SnapSetup) (*gadget.Info, error) {
+	info, err := snap.ReadInfo(snapsup.InstanceName(), snapsup.SideInfo)
+	if err != nil {
+		return nil, err
+	}
+	update, err := snap.ReadGadgetInfo(info, release.OnClassic)
+	if err != nil {
+		return nil, err
+	}
+	return update, nil
+}
+
+func gadgetCurrentAndUpdate(st *state.State, snapsup *snapstate.SnapSetup) (current *gadget.Info, update *gadget.Info, err error) {
+	snapst, err := snapState(st, snapsup.InstanceName())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	currentInfo, err := currentGadgetInfo(snapst)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	newInfo, err := pendingGadgetInfo(snapsup)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return currentInfo, newInfo, nil
+}
+
+var (
+	gadgetUpdate        = nopGadgetOp
+	gadgetRollback      = nopGadgetOp
+	gadgetTrashRollback = nopGadgetTrashRollback
+)
+
+func nopGadgetOp(current, update *gadget.Info, rollbackRootDir string) error {
+	return nil
+}
+
+func nopGadgetTrashRollback(current *gadget.Info, rollbackRootDir string) error {
+	return nil
+}
+
+func (m *DeviceManager) doUpdateGadget(t *state.Task, _ *tomb.Tomb) error {
+	if release.OnClassic {
+		// nothing to do
+		return nil
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	snapsup, err := snapstate.TaskSnapSetup(t)
+	if err != nil {
+		return err
+	}
+	if snapsup.Type != snap.TypeGadget {
+		// not a gadget snap, nothing to update
+		return nil
+	}
+
+	current, update, err := gadgetCurrentAndUpdate(t.State(), snapsup)
+	if err != nil {
+		return err
+	}
+	if current == nil {
+		// XXX: during first boot & seeding?
+		return nil
+	}
+
+	snapRollbackDir, err := makeRollbackDir(snapsup.InstanceName())
```

Comment:
Will there be only one rollback directory? Could we (should we?) support several (e.g. with the revision in the name?)

Reply:
We could include revision in the name for clarity. Although, there's only one snap update at a time, so just name should be sufficiently unique.

Reply:
Discussed with @pedronis and @mvo5, will add snap revision to the directory name.

Reply:
Added